From: Matt Benjamin <mbenjamin@redhat.com>
Date: Fri, 10 Feb 2017 17:14:16 -0500
Subject: rgw_file: add pretty-print for RGWFileHandle

Signed-off-by: Matt Benjamin <mbenjamin@redhat.com>
(cherry picked from commit ef330f385d3407af5f470b5093145f59cc4dcc79)

resolves (in part): rhbz#1420231, rhbz#1416041, rhbz#1415981
---
 src/rgw/rgw_file.cc | 25 +++++++++++++++++++++++++
 src/rgw/rgw_file.h  |  3 +++
 2 files changed, 28 insertions(+)

diff --git a/src/rgw/rgw_file.cc b/src/rgw/rgw_file.cc
index 05cbaec..32e6fc2 100644
--- a/src/rgw/rgw_file.cc
+++ b/src/rgw/rgw_file.cc
@@ -705,6 +705,31 @@ namespace rgw {
     } while (! (stop || shutdown));
   } /* RGWLibFS::gc */
 
+  std::ostream& operator<<(std::ostream &os,
+			   RGWFileHandle const &rgw_fh)
+  {
+    const auto& fhk = rgw_fh.get_key();
+    const auto& fh = const_cast<RGWFileHandle&>(rgw_fh).get_fh();
+    os << "<RGWFileHandle:";
+    os << "addr=" << &rgw_fh;
+    switch (fh->fh_type) {
+    case RGW_FS_TYPE_DIRECTORY:
+	os << "type=DIRECTORY;";
+	break;
+    case RGW_FS_TYPE_FILE:
+	os << "type=FILE;";
+	break;
+    default:
+	os << "type=UNKNOWN;";
+	break;
+      };
+    os << "fid=" << fhk.fh_hk.bucket << ":" << fhk.fh_hk.object << ";";
+    os << "name=" << rgw_fh.object_name() << ";";
+    os << "refcnt=" << rgw_fh.get_refcnt() << ";";
+    os << ">";
+    return os;
+  }
+
   RGWFileHandle::~RGWFileHandle() {
     if (parent && (! parent->is_root())) {
       /* safe because if parent->unref causes its deletion,
diff --git a/src/rgw/rgw_file.h b/src/rgw/rgw_file.h
index bf8d0db..1048748 100644
--- a/src/rgw/rgw_file.h
+++ b/src/rgw/rgw_file.h
@@ -636,6 +636,9 @@ namespace rgw {
 
     virtual ~RGWFileHandle();
 
+    friend std::ostream& operator<<(std::ostream &os,
+				    RGWFileHandle const &rgw_fh);
+
     class Factory : public cohort::lru::ObjectFactory
     {
     public:
